### PR TITLE
Implement token revocation public api

### PR DESCRIPTION
**Pull request by renkelvin**
Implement token revocation public api. Note that it's not e2e tested yet.

Internal API review at [go/gcip-token-revocation-api](http://go/gcip-token-revocation-api)

### CODE REVIEW

**Comment by renkelvin** [comment]
@peterfriese Please take a look, thanks!

File: FirebaseAuth/Sources/Backend/FIRAuthBackend.h [hunk 1]
```diff
@@ -414,8 +425,18 @@ typedef void (^FIRSignInWithGameCenterResponseCallback)(
  */
 + (void)verifyClient:(FIRVerifyClientRequest *)request
             callback:(FIRVerifyClientResponseCallback)callback;
+
 #endif
 
+/** @fn verifyClient:callback:
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/FIRAuthBackend.h:431]: Seems like the doc comments don't match the method signature.

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 Google LLC
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.h:2]: Copyright 2023 Google LLC

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m [hunk 3]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2022 Google LLC
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:2]: Copyright 2023 Google LLC

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m [hunk 4]
```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2022 Google LLC
```
Comment (peterfriese) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m:2]: Copyright 2023 Google LLC

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenRequestTests.m [hunk 5]
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2017 Google
```
Comment (peterfriese) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenRequestTests.m:2]: Copyright 2023 Google LLC

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenResponse.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2022 Google LLC
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenResponse.h:2]: Copyright 2023 Google LLC

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenResponse.m [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2022 Google LLC
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenResponse.m:2]: Copyright 2023 Google LLC

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    @"revokeToken";  // TODO: Double check the endpoint when backend is ready
+
+/** @var kAppTokenKey
+    @brief The key for the appToken request paramenter.
+ */
+static NSString *const kTokenKey = @"token";
+
+/** @var kIDTokenKey
+    @brief The key for the "idToken" value in the request. This is actually the STS Access Token,
+        despite it's confusing (backwards compatiable) parameter name.
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:34]: despite its confusing (backwards compatible) parameter name.

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+static NSString *const kRevokeTokenEndpoint =
+    @"revokeToken";  // TODO: Double check the endpoint when backend is ready
+
+/** @var kAppTokenKey
+    @brief The key for the appToken request paramenter.
+ */
+static NSString *const kTokenKey = @"token";
+
+/** @var kIDTokenKey
+    @brief The key for the "idToken" value in the request. This is actually the STS Access Token,
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:33]: What is STS? Maybe explain this somewhere.

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenRequestTests.m [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+ */
+static NSString *const kFakeToken = @"fakeToken";
+
+/** @var kFakeIDToken
+    @brief The fake ID token to use in the test request.
+ */
+static NSString *const kFakeIDToken = @"fakeIDToken";
+
+/** @var kFakeToken
+    @brief The fake  token to use in the test request.
```
Comment (peterfriese) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenRequestTests.m:38]: remove extra space between "fake" and "token"

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenRequestTests.m [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+#import <XCTest/XCTest.h>
+
+#import "FirebaseAuth/Sources/Backend/FIRAuthBackend.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenResponse.h"
+#import "FirebaseAuth/Tests/Unit/FIRFakeBackendRPCIssuer.h"
+
+/** @var kFakeToken
+    @brief The fake  token to use in the test request.
```
Comment (peterfriese) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenRequestTests.m:28]: Extra space between "fake" and "token"

**Review by peterfriese — CHANGES_REQUESTED** [review]
Some minor documentation issues.

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+ */
+static NSString *const kTokenKey = @"token";
+
+/** @var kIDTokenKey
+    @brief The key for the ID Token associated with this credential.
+ */
+static NSString *const kIDTokenKey = @"idToken";
+
+typedef NS_ENUM(NSInteger, FIRTokenType) {
+  /** Indicates that the token type is uspecified.
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:48]: unspecified

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h [hunk 13]
```diff
@@ -850,6 +850,16 @@ NS_SWIFT_NAME(Auth)
  */
 - (BOOL)canHandleNotification:(NSDictionary *)userInfo API_UNAVAILABLE(macos, tvos, watchos);
 
+/** @fn revokeToken:Completion
+    @brief Revoke the given token (Currently only supports Apple issued access token, refresh token
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h:854]: "Currently|" should be lowercase: "currently"

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h [hunk 14]
```diff
@@ -850,6 +850,16 @@ NS_SWIFT_NAME(Auth)
...
+   and authorization code).
+
+    @param completion Optionally; the block invoked when the request to revoke the token is
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h:857]: (Optional) the block invoked

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+#import "FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuthErrors.h"
+
+#import "FirebaseAuth/Sources/Backend/FIRAuthBackend.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.h"
+#import "FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenResponse.h"
+#import "FirebaseAuth/Tests/Unit/FIRFakeBackendRPCIssuer.h"
+
+/** @var kFakeToken
+    @brief The fake  token to use in the test request.
```
Comment (peterfriese) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m:30]: There is an extra space between fake and token

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+ */
+static NSString *const kFakeToken = @"fakeToken";
+
+/** @var kFakeIDToken
+    @brief The fake ID token to use in the test request.
+ */
+static NSString *const kFakeIDToken = @"fakeIDToken";
+
+/** @var kFakeToken
+    @brief The fake  token to use in the test request.
```
Comment (peterfriese) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m:40]: Same here: extra space.

**Review by peterfriese — COMMENTED** [review]
LGTM, just a couple of nits.

File: FirebaseAuth/Sources/Auth/FIRAuth.m [hunk 17]
```diff
@@ -1537,6 +1539,30 @@ - (void)setAdditionalFrameworkMarker:(nullable NSString *)additionalFrameworkMar
   });
 }
 
+- (void)revokeToken:(NSString *)token
+         completion:(nullable void (^)(NSError *_Nullable error))completion {
+  [self.currentUser
+      getIDTokenWithCompletion:^(NSString *_Nullable idToken, NSError *_Nullable error) {
+        if (error) {
```
Comment (paulb777) [inline FirebaseAuth/Sources/Auth/FIRAuth.m:1546]: Missing ` if (completion) {` here and below.

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h [hunk 18]
```diff
@@ -850,6 +850,16 @@ NS_SWIFT_NAME(Auth)
  */
 - (BOOL)canHandleNotification:(NSDictionary *)userInfo API_UNAVAILABLE(macos, tvos, watchos);
 
+/** @fn revokeToken:Completion
+    @brief Revoke the given token (currently only supports Apple issued access token, refresh token
```
Comment (paulb777) [inline FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h:854]: The API review claims all tokens are supported.

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+@implementation FIRRevokeTokenRequest
+
+- (nullable instancetype)initWithToken:(NSString *)token
+                               idToken:(NSString *)idToken
+                  requestConfiguration:(FIRAuthRequestConfiguration *)requestConfiguration {
+  self = [super initWithEndpoint:kRevokeTokenEndpoint requestConfiguration:requestConfiguration];
+  if (self) {
+    // Apple and Authorization code are the only provider and token type we support for now.
+    _providerID = @"apple.com";
```
Comment (paulb777) [inline FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:73]: Since the plan is to support other providers, shouldn't `providerID` be a parameter?

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+           RPCError = error;
+           callbackInvoked = YES;
+         }];
+
+  [_RPCIssuer respondWithJSON:@{}];
+
+  XCTAssert(callbackInvoked);
+  XCTAssertNotNil(RPCResponse);
+}
+
```
Comment (paulb777) [inline FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m:114]: Add a test for a nil callback

**Review by paulb777 — COMMENTED** [review]
Please add build tests for the new API in https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseAuth/Tests/Sample/SwiftApiTests/SwiftAPI.swift

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m (hunk 20 above)
Comment (renkelvin) [reply FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m:114]: The callback is nonnull. You mean a nil response?

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m (hunk 19 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:73]: The plan is, while our backend only supports Apple for now, so it's limited to Apple to be less error-prone. WDYT?

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h (hunk 18 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h:854]: Thanks for catching this.
Technically, our backend accepts all 3 types of tokens, however on Apple platforms, the developer only gets authorization code (refresh token and access token will be returned in Web and Android flows).
I suggest we should change the comments here. Let me confirm with the api council.

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m (hunk 19 above)
Comment (paulb777) [reply FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:73]: Since it's an internal API, I don't feel strongly either way

File: FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m (hunk 20 above)
Comment (paulb777) [reply FirebaseAuth/Tests/Unit/FIRRevokeTokenResponseTests.m:114]: The second parameter to revokeToken is nullable. See line 1543 of FIRAuth.m

File: FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m (hunk 19 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/Backend/RPC/FIRRevokeTokenRequest.m:73]: Added a comment to generalize the initializer in the future. Will keep the api as is.

File: FirebaseAuth/Sources/Backend/FIRAuthBackend.h [hunk 21]
```diff
@@ -220,6 +222,15 @@ typedef void (^FIRVerifyPhoneNumberResponseCallback)(
 typedef void (^FIRVerifyClientResponseCallback)(FIRVerifyClientResponse *_Nullable response,
                                                 NSError *_Nullable error);
 
+/** @typedef FIRVerifyClientResponseCallback
+    @brief The type of block used to return the result of a call to the verifyClient endpoint.
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/FIRAuthBackend.h]: Update the docs to match the method signature.

File: FirebaseAuth/Sources/Backend/FIRAuthBackend.h [hunk 22]
```diff
@@ -578,8 +599,18 @@ typedef void (^FIRSignInWithGameCenterResponseCallback)(
  */
 - (void)verifyClient:(FIRVerifyClientRequest *)request
             callback:(FIRVerifyClientResponseCallback)callback;
+
 #endif
 
+/** @fn verifyClient:callback:
+    @brief Calls the verifyClient endpoint, which is responsible for sending the silent push
+        notification used for app validation to the device provided in the request parameters.
```
Comment (peterfriese) [inline FirebaseAuth/Sources/Backend/FIRAuthBackend.h]: Update the docs to match the method signature.

**Review by peterfriese — CHANGES_REQUESTED** [review]

**Review by peterfriese — APPROVED** [review]

File: FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h (hunk 18 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuth.h]: Confirmed with the api council that it's ok with changing the comments here.

**Comment by renkelvin** [comment]
Due to backend limitation, we gonna revoke refresh token instead of authorization code. Please help review the PR again. Thanks!

**Review by paulb777 — COMMENTED** [review]
Please fix CI. You may need to rebase or merge master to do so.

**Comment by renkelvin** [comment]
> Please fix CI. You may need to rebase or merge master to do so.

Interesting, I think I got a green CI initially..
The diff messed up after rebasing. Recreated https://github.com/firebase/firebase-ios-sdk/pull/11001
